src/errors.rs

```diff
@@ -97,6 +97,9 @@ pub enum R1CSError {
     /// Occurs when there are insufficient generators for the proof.
     #[fail(display = "Invalid generators size, too few generators for proof")]
     InvalidGeneratorsLength,
+    /// This error occurs when the proof encoding is malformed.
+    #[fail(display = "Proof data could not be parsed.")]
+    FormatError,
```

Comment:
Since we do not decompress points or check scalars until much later in the verification code, the FormatError is simply an error in size of a proof. IMHO it does not provide useful diagnostic information, so we can keep using the plain "ProofError" like elsewhere in the code. Also, users cannot really recover from this error, so if the proof is invalid (for whatever reason), the whole user's protocol should be aborted the same way (cf. recent [PGP efail attack](https://efail.de)).

Reply:
I felt like `ProofError` was `RangeProof`-specific, but it indeed felt awkward to add this. Will add a separate commit, and if it's indeed better I'll `fixup` the first commit.

Reply:
> you now would need to convert from both ProofError and R1CSError to your own Error format if you're writing a program.

This seems to be the case no matter what error types we use in-library, because if the user has a different error type, they'd have to convert or wrap this one anyway.

Reply:
(deleted previously posted incorrect comment)

I don't think you'd often have a code that uses both RP and R1CS APIs. Because if you already use R1CS, then it's cheaper to just use rangeproof-as-a-gadget within R1CS, w/o using a dedicated rangeproof API at all.

Reply:
Indeed my point. `ProofError` has `RangeProof`-specific values, so if the deserialization of `R1CSProof` would return a `ProofError`, a R1CS user would need to handle the `RangeProof`-error values, which doesn't make sense.

Do you still want me to return `ProofError`?

Reply:
Oh, I apologize for confusion I caused. Originally I meant R1CSError when i said ProofError.